icu4j/main/core/src/main/java/com/ibm/icu/segmenter/SegmentsImpl.java

```diff
@@ -0,0 +1,128 @@
+// © 2025 and later: Unicode, Inc. and others.
+// License & terms of use: https://www.unicode.org/copyright.html
+
+package com.ibm.icu.segmenter;
+
+import com.ibm.icu.text.BreakIterator;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+class SegmentsImpl implements Segments {
+
+  private CharSequence source;
+
+  private BreakIterator breakIterPrototype;
+
+  SegmentsImpl(BreakIterator breakIter, CharSequence source) {
+    this.source = source;
+
+    // We are creating a clone of the Segmenter's prototype BreakIterator field so that this
+    // concrete Segments object can avoid sharing state with the other Segments object instances
+    // that get spawned from the Segmenter. This allows difference source CharSequences to be used
+    // in each Segments object.
+    //
+    // In turn, the cloned BreakIterator becomes a prototype to be stored in the Segments object,
+    // which then gets cloned and used in each of the Segments APIs' implementations. The second
+    // level of cloning that happens when the Segments object's local BreakIterator prototype
+    // gets cloned allows the iteration state to be separate whenever an Segments API is called.
+    // Otherwise, there is a chance that multiple API calls on the same Segments object might
+    // mutate the same position/index, if done concurrently.
+    breakIterPrototype = (BreakIterator) breakIter.clone();
+    // It's okay to perform .setText on the object that we want to clone later because we should
+    // then not have to call .setText on the clones.
+    breakIterPrototype.setText(source);
+  }
+
+  @Override
+  public Segment segmentAt(int i) {
+    BreakIterator breakIter = (BreakIterator) breakIterPrototype.clone();
+    int start;
+    int limit;
+
+    boolean isBoundary = breakIter.isBoundary(i);
+
+    if (isBoundary) {
+      start = i;
+      limit = breakIter.next();
+    } else {
+      // BreakIterator::isBoundary(i) will advance forwards to the next boundary if the argument
+      // is not a boundary.
+      limit = breakIter.current();
+      start = breakIter.previous();
+    }
+
+    if (start != BreakIterator.DONE && limit != BreakIterator.DONE) {
+      return new Segment(start, limit, source);
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  public Stream<Segment> segments() {
```

Comment:
call segmentsFrom(0) to reduce code?

maybe even do so in the interface definition?